smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/ResolveEndpointMiddlewareGenerator.kt

```diff
@@ -0,0 +1,153 @@
+        renderClassMembers()
+        writer.write("")
+        renderInstall()
+        writer.closeBlock("}")
+    }
+
+    private fun renderConstructorParams() {
+        writer.write("private val endpointProvider: #T,", EndpointProviderGenerator.getSymbol(ctx.settings))
+        writer.write("private val buildParams: #T.Builder.(input: I) -> Unit,", EndpointParametersGenerator.getSymbol(ctx.settings))
+    }
+
+    private fun renderClassMembers() {
+        writer.write("private lateinit var endpoint: #T", RuntimeTypes.Http.Endpoints.Endpoint)
+    }
+
+    private fun renderInstall() {
+        writer.withBlock("override fun install(op: #T<I, O>) {", "}", RuntimeTypes.Http.Operation.SdkHttpOperation) {
+            renderInitialize()
+            write("")
+            renderMutate()
+        }
+    }
+
+    private fun renderInitialize() {
+        writer.withBlock("op.execution.initialize.intercept { req, next ->", "}") {
+            write("val params = #T { buildParams(req.subject) }", EndpointParametersGenerator.getSymbol(ctx.settings))
+            write("endpoint = endpointProvider.resolveEndpoint(params)")
+            write("next.call(req)")
+        }
+    }
+
+    private fun renderMutate() {
+        writer.withBlock("op.execution.mutate.intercept { req, next ->", "}") {
+            write("#T(req, endpoint)", RuntimeTypes.Http.Endpoints.setResolvedEndpoint)
+            renderPostResolution()
+            write("next.call(req)")
+        }
+    }
+}
+
+/**
+ * Renders an endpoint parameter builder lambda that binds values for a specific operation.
+ */
+class EndpointParameterBindingGenerator(
+    private val model: Model,
+    private val service: ServiceShape,
+    private val writer: KotlinWriter,
+    private val op: OperationShape,
+    private val rules: EndpointRuleSet,
+    private val inputPrefix: String,
+) {
+    private val opIndex = OperationIndex.of(model)
+    private val staticContextParams = op.getTrait<StaticContextParamsTrait>()
+
+    // maps endpoint parameter name -> input member shape
+    private val inputContextParams = buildMap<String, MemberShape> {
+        opIndex.getInput(op).getOrNull()?.members()?.forEach { member ->
+            member.getTrait<ContextParamTrait>()?.let { trait ->
+                put(trait.name, member)
+            }
+        }
+    }
+    private val clientContextParams = service.getTrait<ClientContextParamsTrait>()
+
+    fun render() {
+        rules.parameters.toList().forEach(::renderBinding)
+    }
+
+    /**
+     * Render the assignment of an endpoint param value within a builder. The SEP dictates a specific source order to
+     * use when determining how to bind:
+     * 1. staticContextParams (from operation shape)
+     * 2. contextParam (from member of operation input shape)
+     * 3. clientContextParams (from service shape)
+     * 4. builtin binding
+     * 5. builtin default
```

Comment:
Do we have any instances of (5)? I'm assuming `bindAwsBuiltins` is type (4) and why we render it _before_ delegating to `EndpointParameterBindingGenerator` in `aws-sdk-kotlin`?

Reply:
Correct, and (5) is essentially fulfilled by the config field defaults.